Add signedEntryTimestamp to LogEntry for offline verification

priyawadhwa opened the pull request:
This PR adds a signature field to LogEntry (I'm open to different names for signature here since it might be overused). It can be used to verify the LogEntry was signed for offline verification, which cosign will be using for bundling!

Basically, rekor will:
1. Sign the LogEntry
2. Add the signature to the "Verification.SignedEntryTimestamp" field of the LogEntry
3. Send that to the client whenever they create a new entry

and clients will:
1. Get the LogEntry when they create a new entry, and set signature=nil to get the payload (making people set signature=nil is kinda weird, but follows the precert model where the poison extension has to be removed to validate the cert). I went this route so that the signature could be bundled with the LogEntry which is already being returned when users add an entry to the log.
2. Validate the payload and signature against the rekor public key

Also adds an integration test for this.

cc @dlorenc @asraa

Commit:
Add signature to LogEntry for offline verification

Also add an integration test for this.

Signed-off-by: Priya Wadhwa <priyawadhwa@google.com>

dlorenc commented on openapi.yaml:
```diff
@@ -301,6 +301,10 @@ definitions:
           type: integer
         inclusionProof:
           $ref: '#/definitions/InclusionProof'
+        signature:
```
we should probably be a bit more verbose with the name, `signedEntryTimestamp` maybe?

asraa replied:
I almost feel like log entry should be a type in itself with logIndex, body, and integrated time so it's clear log entry includes those three (and not e.g. inclusion proof)

asraa replied:
btw, do we want this to be replaced with an RFC 3161 signed timestamp over (logIndex, body) in the future?

dlorenc replied:
Probably not replaced by, but in addition to? I don't love the idea of promoting usage of the RFC3161 format outside of where we need it for compatibility.

asraa replied:
Oh yeah, definitely. the other option is a JSON based one, which would look roughly like a packaged signature field + message digest + cert chain

bobcallaway replied:
I'm also in favor of shifting the inclusion proof and the signature into their own level of the LogEntry object. We should also probably add the log ID (in case we ever needed to rotate to a different log for performance / maintenance reasons)

perhaps something like:
```
LogEntry:
    type: object
    additionalProperties:
      type: object
      properties:
        logId:
          type: string
        logIndex:
          type: integer
          minimum: 0
        body:
          type: object
          additionalProperties: true
        integratedTime:
          type: integer
        verification:
          type: object
          properties:
            inclusionProof:
              $ref: '#/definitions/InclusionProof'
            signedEntryTimestamp:
              $ref: '#/definitions/SignedEntryTimestamp'
      required:
        - "logId"
        - "logIndex"
        - "body"
        - "integratedTime"
```

priyawadhwa replied:
Cool, I renamed it to `signedEntryTimestamp` and moved `signedEntryTimestamp` and `inclusionProof` under `verification`

I figure we can add in the logID in a separate PR so that this one doesn't get too big!

dlorenc commented on tests/e2e_test.go:
```diff
@@ -359,3 +374,117 @@ func TestWatch(t *testing.T) {
+	}
+	// unmarshal the response
+	le := resp.GetPayload()
+	if len(le) != 1 {
```
we should make a helper for this at some point (getting the one field out), it's so annoying to get right and we do it all over the place

priyawadhwa replied:
I added it as a helper func for testing, but if we're doing it in code I could move it to a general util package or something

dlorenc reviewed (commented):
This is awesome, a couple nits then I want @bobcallaway to check the API change

bobcallaway commented on openapi.yaml:
```diff
@@ -301,6 +301,10 @@ definitions:
+        signature:
+          type: string
+          format: byte
+          description: signature over the log entry (acts as an inclusion promise if inclusion proof is not included)
```
right now Rekor is blocking on the insertion loop so this wouldn't really be a promise... but more of a static attestation of inclusion. Not sure how much to stress over the semantics here.

dlorenc replied:
Right, I was thinking we might want to leave it open for interpretation later in case we ever need to decouple these. I'm fine keeping it simple though.

priyawadhwa replied:
sgtm, I made the description more general here

bobcallaway commented on pkg/api/entries.go:
```diff
@@ -156,6 +155,21 @@ func CreateLogEntryHandler(params entries.CreateLogEntryParams) middleware.Respo
 		}()
 	}
 
+	payload, err := logEntryAnon.MarshalBinary()
```
This is just marshaling to JSON under the covers... we should probably marshal it in a JCS-compatible way before signing so that it is repeatable for clients to verify the hash.  

Also since we support returning data in YAML format I think we need to be explicit in documenting how we format the data before signing since there would be a potential for mismatch there too.

priyawadhwa replied:
Cool, I found this json canonicalization library & added it in: https://github.com/cyberphone/json-canonicalization

Where would be the best spot for the documentation? I could make a `docs` folder in the repo if there's isn't an existing spot for this kind of thing.

dlorenc replied:
Nothing really right now outside of the openapi.yaml. I'd try to be as verbose there as you can and then see if we need more space.

priyawadhwa replied:
cool, I added in instructions for how to verify as comments

Commit:
Address code review comments:
- Canonicalize payload before signing it
- Change name of signature to signedEntryTimestamp
- move signedEntryTimestamp and inclusionProof into separate Verification field in LogEntry

Signed-off-by: Priya Wadhwa <priyawadhwa@google.com>

Commit:
Create helper func for extracting log entry

Signed-off-by: Priya Wadhwa <priyawadhwa@google.com>

Commit:
Add info around verifying signedEntryTimestamp as comments in openapi.yaml

Signed-off-by: Priya Wadhwa <priyawadhwa@google.com>

dlorenc reviewed (commented):
Nice, LGTM now.

@asraa and @bobcallaway any other thoughts?

bobcallaway commented on openapi.yaml:
```diff
@@ -299,8 +299,20 @@ definitions:
+            signedEntryTimestamp:
+              type: string
+              format: byte
+              # To verify the signedEntryTimestamp:
```
if you end up making another change to this, I might add change the wording to be a bit less Golang specific - for example:

- Remove the verification object from the JSON document
- Follow RFC 8785 rules to canonicalize the remaining JSON document

priyawadhwa replied:
good call, added that in

asraa reviewed (approved):
LGTM!

Commit:
Generalize verification instructions

Signed-off-by: Priya Wadhwa <priyawadhwa@google.com>
